### DISCUSSION TITLE
CefSharp browser jumps (scrolls up) when "PrintToPdfAsync" method is called

### DISCUSSION THREAD

**Discussion by hoquras** [cefsharp/CefSharp#4115] [General]
I am using the "PrintToPdfAsync" method [https://cefsharp.github.io/api/63.0.0/html/M_CefSharp_WebBrowserExtensions_PrintToPdfAsync.htm](url) to export HTML view to pdf. However I am experiencing following issues:

1. With this API (PrintToPdfAsync), I am unable to pass complete "viewport" that I want to be exported. I can pass width and height (possibly page dimension in micron). Question: How cefSharp determines the "view port"? Which element window or body?

3. The major issue I am experiencing is that just after the call to the method, "view is scrolled up", thereby the exported pdf does not contain the expected view port area. Here is a screen capture: 

![JewelSuite_8zY4ZUyCFf](https://user-images.githubusercontent.com/105909036/169481982-6ef9a13f-5e7e-48e1-b709-2135e301ede4.gif)

So the exported pdf is contains area (top part) that was not part of the intended export.
![image](https://user-images.githubusercontent.com/105909036/169482254-6543156b-c9b1-4f60-a6e4-b5642ceb62e5.png)

**Comment by amaitland** [comment]
> I am using the "PrintToPdfAsync" method [https://cefsharp.github.io/api/63.0.0/html/M_CefSharp_WebBrowserExtensions_PrintToPdfAsync.htm](https://github.com/cefsharp/CefSharp/discussions/url) to export HTML view to pdf.

Firstly the API doc is version specific http://cefsharp.github.io/api/
You are referencing version 63 which is very old.

>  I can pass width and height (possibly page dimension in micron).

There are no width/height properties, there are PageWidth/PageHeight which are intended to specify the page size, A4, A5, letter etc.

> The major issue I am experiencing is that just after the call to the method, "view is scrolled up", thereby the exported pdf does not contain the expected view port area. 

Testing with version 101.0.180 using WinForms, default settings and I'm not seeing any jump. How does the default page size behave? 

Rendering and PDF generation are all implemented at the Chromium level, make sure you are using the latest version.

**Reply by hoquras** [reply]
Dear Alex, Many thanks for your reply.

I have also tested with the latest 101.0.180. I can see the jump. I have made a page with 3 simple divs (parent div contains scrollbar). The code is pretty simple:
`<div
        style={{
          display: 'flex',
          height: '100vh',
          border: '1px solid red',
          overflow: 'auto',
          padding: '5px',
        }}
      >
        <div
          style={{
            width: 200,
            height: 2000,
            color: 'blue',
            border: '1px solid blue',
            margin: '10px',
          }}
        >
...`
And here is what I see:
![CefSharp MinimalExample WinForms_u8ajX7J9va](https://user-images.githubusercontent.com/105909036/169920907-3edd5379-a9de-4597-9a78-5039e53268bf.gif)

If you see the end, the scrollbar jumped up and exported pdf captured the top part of the view which was not on the viewport.

I did not pass the default setting. Here the intention is **not** to export the whole page but only the visible part that is the "view port". Therefore, I am setting "ClientHeight" as the "PageHeight" and "ClientWidth" as the "PageWidth" (converting pixel to micron). It seems that when zoom is 100% it works (the browser correctly identifies the view port position). However when zoom is around 40%, the browser identification wrong view port position. 

Wondering with with 100% zoom, even if we scroll the end of the page, browser determines "view port" correctly. However with zoom level 40% or less, "view port" is possibly not correctly calculated. How does Cef determines the view port? Is there any relation with zoom level?

**Reply by hoquras** [reply]
May be the important question, is "PrintToPdfAsync" right API to export a partial page (scrollable) i.e "viewport" to pdf? The api does not let us to pass "view port".

**Reply by amaitland** [reply]
> How does Cef determines the view port? Is there any relation with zoom level?

The question should be how does `Chromium`? I'm not sure in this context, you'd have to look at the source.

> May be the important question, is "PrintToPdfAsync" right API to export a partial page (scrollable) i.e "viewport" to pdf? The api does not let us to pass "view port".

No. There's no such `API` that I'm aware of. `Chromium` expresses `PDFs` as pages.

**Reply by hoquras** [reply]
Thanks for your reply.

**Reply by amaitland** [reply]
CEF doesn't currently expose all the functionality that `Chromium` has to offer. 

https://github.com/chromiumembedded/cef/blob/cb83a7c3fddacb984df60d303701a3d25ad1f826/libcef/browser/printing/print_view_manager.cc#L60
There are other settings like [PageRange](https://source.chromium.org/chromium/chromium/src/+/main:printing/print_job_constants.h;l=80?q=kSettingRasterizePdf&ss=chromium) that can give finer control of the output.

You can capture a screenshot for a given rect.

**Reply by hoquras** [reply]
Yes I understand now. The issue here is that screenshot won't work. We are exporting high quality SVG image to pdf. Screenshot won't help, I think.

Thanks again. I highly appreciate your comments.
Regards-

**Reply by amaitland** [reply]
Perhaps something like https://github.com/dlemstra/Magick.NET would be better suited to converting SVG to PDF.
